rust/kernel/module_param.rs

```diff
@@ -300,3 +323,152 @@ make_param_ops!(
     PARAM_OPS_BOOL,
     bool
 );
+
+/// An array of at __most__ `N` values.
+pub struct ArrayParam<T, const N: usize> {
+    values: [core::mem::MaybeUninit<T>; N],
+    used: usize,
+}
+
+impl<T, const N: usize> ArrayParam<T, { N }> {
+    fn values(&self) -> &[T] {
+        // SAFETY: `self` can only be generated and modified by the methods
+        // [`new`], and [`push`]. These maintain the invariant that the first
+        // `self.used` elements of `self.values` are initialized.
+        unsafe {
+            &*(&self.values[0..self.used] as *const [core::mem::MaybeUninit<T>] as *const [T])
+        }
+    }
+}
+
+impl<T: Copy, const N: usize> ArrayParam<T, { N }> {
+    const fn new() -> Self {
+        // Invariant:
+        // The first `self.used` elements of `self.values` are initialized.
```

Comment:
Let's match `SAFETY` comments case:

```suggestion
        // INVARIANT: the first `self.used` elements of `self.values` are initialized.
```

i.e. this makes clear this is a piece of code that works to uphold the invariant defined elsewhere as `# Invariant`; akin to `# Safety` docs vs. `SAFETY` blocks.